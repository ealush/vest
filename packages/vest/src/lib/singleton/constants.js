/**
 * @type {String} Vest's major version.
 */
const VEST_MAJOR = VEST_VERSION.split('.')[0];

/**
 * @type {Symbol} Used to store a global reference to Vest's inner state.
 */
export const SYMBOL_VEST_GLOBAL = Symbol.for(`__INTERNAL__VEST#${VEST_MAJOR}`);
