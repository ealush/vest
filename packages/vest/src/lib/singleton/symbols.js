/**
 * @type {String} Vest's major version.
 */
const VEST_MAJOR = VEST_VERSION.split('.')[0];

/**
 * @type {Symbol} Used to store a global instance of Vest.
 */
export const SYMBOL_VEST = Symbol.for(`VEST#${VEST_MAJOR}`);
