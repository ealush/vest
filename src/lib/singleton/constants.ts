/**
 * Vest's major version.
 */
// @ts-ignore
const VEST_MAJOR: string = VEST_VERSION.split('.')[0];

/**
 * Used to store a global instance of Vest.
 */
export const SYMBOL_VEST: symbol = Symbol.for(`VEST#${VEST_MAJOR}`);
