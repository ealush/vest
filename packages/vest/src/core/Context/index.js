import singleton from '../../lib/singleton';

/**
 * Creates a new context object, and assigns it as a static property on Vest's singleton.
 * @param {Object} parent   Parent context.
 * @returns {Context} either an existing or a new context object.
 */

class Context {
  /**
   * Clears stored instance from constructor function.
   */
  static clear() {
    const ctx = singleton.useContext();

    if (ctx?.parentContext) {
      singleton.use().ctx = ctx.parentContext;
      ctx.parentContext.removeChildContext();
    } else {
      singleton.use().ctx = null;
    }
  }

  constructor(ctxRef) {
    const ctx = singleton.useContext();
    Object.assign(this, ctxRef);

    if (ctx) {
      ctx.setChildContext(this);
    }

    singleton.use().ctx = this;
  }

  get suiteId() {
    return this.lookup('suite_id');
  }

  set suiteId(suiteId) {
    this.suite_id = suiteId;
  }

  lookup(key) {
    let ctx = this;
    do {
      if (ctx[key]) {
        return ctx[key];
      }
      ctx = ctx.parentContext;
    } while (ctx || ctx?.parentContext);
  }

  setParentContext(parentContext) {
    this.parentContext = parentContext;
  }

  setChildContext(childContext) {
    childContext.parentContext = this;
    this.childContext = childContext;
  }

  removeChildContext() {
    this.childContext = null;
  }

  /**
   * Sets a testObject reference on context.
   * @param {VestTest} A VestTest instance.
   */
  setCurrentTest(testObject) {
    this.currentTest = testObject;
  }

  /**
   * Removes current test from context.
   */
  removeCurrentTest() {
    delete this.currentTest;
  }
}

export default Context;
