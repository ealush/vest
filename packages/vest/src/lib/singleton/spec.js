import faker from 'faker';
import _ from 'lodash';
import vest from '../..';
import { version } from '../../../package.json';
import mock from '../../../testUtils/mock';
import { ALL_VEST_BUILDS } from '../../../testUtils/vestBuilds';
import state from '../../core/state';
import go from '../globalObject';
import { SYMBOL_VEST_GLOBAL } from './constants';
import singleton from '.';

describe('singleton', () => {
  afterAll(() => {
    go.VEST_VERSION = vest.VERSION;
    singleton.register();
  });

  describe('Attaching to global scope', () => {
    beforeEach(() => {
      delete go[SYMBOL_VEST_GLOBAL];
    });

    afterEach(() => {
      delete go[SYMBOL_VEST_GLOBAL];
    });

    test('Santiy: That global instance is not populated yet', () => {
      expect(go[SYMBOL_VEST_GLOBAL]).toBeUndefined();
    });

    describe('When already registered', () => {
      describe('When same version', () => {
        it('Should return silently', () =>
          new Promise(done => {
            const timeout = setTimeout(() => done(), 300);

            process.on('uncaughtException', () => {
              clearTimeout(timeout);
            });
          }));
      });

      describe('When different version', () => {
        let singleton, mockThrowError;

        beforeEach(() => {
          mockThrowError = mock('throwError');
          singleton = require('.');
        });

        afterEach(() => {
          jest.resetAllMocks();
        });

        it('Should throw an error', () => {
          global.VEST_VERSION = '222';
          singleton.register();
          expect(mockThrowError.mock.calls[0][0]).toBe(
            `Multiple versions of Vest detected: (222,${version}).\n    Most features should work regularly, but for optimal feature compatibility, you should have all running instances use the same version.`
          );
          global.VEST_VERSION = vest.VERSION;
        });
      });
    });
  });
});
describe('Integration: singleton', () => {
  beforeAll(() => {
    singleton.register();
    state.register();
  });

  const pairs = ALL_VEST_BUILDS.reduce(
    (pairs, { validate }) => [
      ...pairs,
      ...ALL_VEST_BUILDS.map(({ test }) => [validate, test]),
    ],
    []
  );

  it.each(pairs)(
    'Should produce correct validation result with `test` and `validate` from different builds',
    (validate, test) => {
      const errorCount = _.random(1, 10);
      const successCount = _.random(1, 10);
      const warnCount = _.random(1, 10);
      const output = validate(faker.random.word(), () => {
        Array.from({ length: warnCount }, () =>
          test(faker.random.word(), faker.lorem.sentence(), function () {
            this.warn();
            return false;
          })
        );
        Array.from({ length: errorCount }, () =>
          test(faker.random.word(), faker.lorem.sentence(), () => false)
        );
        Array.from({ length: successCount }, () =>
          test(faker.random.word(), faker.lorem.sentence(), () => true)
        );
      });

      expect(output.errorCount).toBe(errorCount);
      expect(output.warnCount).toBe(warnCount);
    }
  );
});
