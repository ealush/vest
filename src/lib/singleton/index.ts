import throwError from '../throwError';
import go from '../globalObject';
import { SYMBOL_VEST } from './constants';

/**
 * Throws an error when multiple versions of Vest are detected on the same runtime.
 * @param  {String[]} versions List of detected Vest versions.
 */
const throwMultipleVestError = (...versions) => {
    throwError(`Multiple versions of Vest detected: (${versions.join()}).
    Most features should work regularly, but for optimal feature compatibility, you should have all running instances use the same version.`);
};

/**
 * Registers current Vest instance on global object.
 * @param {Object} vest Reference to Vest.
 * @return {Function} Global Vest reference.
 */
const register = (vest) => {

    const existing = go[SYMBOL_VEST];

    if (existing) {
        if (existing.VERSION !== vest.VERSION) {
            throwMultipleVestError(vest.VERSION, existing.VERSION);
        }
    } else {
        go[SYMBOL_VEST] = vest;
    }

    return go[SYMBOL_VEST];
};

/**
 * @returns Global Vest instance.
 */
const use = () => go[SYMBOL_VEST];

/**
 * @returns Current Vest context.
 */
const useContext = () => use().ctx;

export default {
    use,
    useContext,
    register
};
