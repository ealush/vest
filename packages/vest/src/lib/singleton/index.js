import go from '../globalObject';
import throwError from '../throwError';
import { SYMBOL_VEST } from './symbols';
/**
 * Throws an error when multiple versions of Vest are detected on the same runtime.
 * @param  {String[]} versions List of detected Vest versions.
 */
const throwMultipleVestError = (...versions) => {
  throwError(`Multiple versions of Vest detected: (${versions.join()}).
    Most features should work regularly, but for optimal feature compatibility, you should have all running instances use the same version.`);
};

/**
 * Registers current Vest instance on global object.
 * @param {Object} vest Reference to Vest.
 * @return {Function} Global Vest reference.
 */
const register = (vest, cb) => {
  const existing = go[SYMBOL_VEST];

  if (existing) {
    if (existing.VERSION !== vest.VERSION) {
      throwMultipleVestError(vest.VERSION, existing.VERSION);
    }
  } else {
    go[SYMBOL_VEST] = vest;
  }

  cb();

  return go[SYMBOL_VEST];
};

/**
 * @returns Global Vest instance or a portion of it.
 */
const use = key => (key ? go[SYMBOL_VEST][key] : go[SYMBOL_VEST]);

/**
 * Updates a key in the state.
 * @param {string} key
 * @param {*} value
 */
const set = (key, value) => (go[SYMBOL_VEST][key] = value);

/**
 * @returns Current Vest context.
 */
const useContext = () => use().ctx;

export default {
  use,
  set,
  useContext,
  register,
};
