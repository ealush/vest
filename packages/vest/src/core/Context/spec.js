import faker from 'faker';
import singleton from '../../lib/singleton';
import Context from '.';

const resetContext = () => {
  do {
    Context.clear();
  } while (singleton.useContext());
};

describe('Context', () => {
  let parent, instance;

  beforeEach(() => {
    parent = {
      [faker.random.word()]: faker.random.word(),
      [faker.random.word()]: faker.random.word(),
      [faker.random.word()]: faker.random.word(),
    };

    instance = new Context(parent);
  });

  afterEach(() => {
    Context.clear();
  });

  it('Should assign all parent properties onto ctx instance', () => {
    Object.keys(parent).forEach(key => {
      expect(instance[key]).toBe(parent[key]);
    });
  });

  it('Should store instance on singleton', () => {
    expect(singleton.useContext()).toBe(instance);
  });

  describe('Context.clear', () => {
    describe('When no nested context', () => {
      it('Should nullify stored instance', () => {
        expect(singleton.useContext()).toBe(instance);
        Context.clear();
        expect(singleton.useContext()).toBeNull();
      });
    });

    describe('When nested context', () => {
      let contextInstances;
      beforeEach(() => {
        resetContext();
        contextInstances = Array.from({ length: 10 }, () => new Context());
      });

      it('Should set current context to context parent and nullify its context child', () => {
        // sanity
        expect(contextInstances).toHaveLength(10);
        let counter = 0;

        contextInstances.reverse().forEach(instance => {
          expect(singleton.useContext()).toBe(instance);
          const parent = instance.parentContext;

          Context.clear();

          if (parent) {
            counter++;
            expect(singleton.useContext()).toBe(parent);
          } else {
            expect(singleton.useContext()).toBeNull();
          }
        });

        // sanity
        expect(counter).toBe(9);
      });
    });
  });

  describe('getter: suiteId', () => {
    let suiteId;

    beforeEach(() => {
      Context.clear();
      suiteId = faker.random.uuid();
    });

    describe('When suiteId is present on current context', () => {
      it('Should return suiteId', () => {
        expect(singleton.useContext()?.suiteId).toBeUndefined();
        new Context({ suiteId });
        expect(singleton.useContext().suiteId).toBe(suiteId);
      });
    });

    describe('When suite id is higher in the parent tree', () => {
      it('Should get closest suiteId', () => {
        new Context({ suiteId: 'not_closest' });
        new Context({ suiteId });
        new Context({});
        new Context({});
        expect(singleton.useContext().suiteId).toBe(suiteId);
      });
    });
  });

  describe('setChildContext', () => {
    let context;
    beforeEach(() => {
      resetContext();
      context = new Context();
    });

    it('Should set childContext on current context instance', () => {
      const childContext = {};

      expect(context.childContext).toBeUndefined();
      context.setChildContext(childContext);
      expect(context.childContext).toBe(childContext);
    });

    it('Should set currentContext as childContext parent', () => {
      const childContext = {};

      expect(childContext.parentContext).toBeUndefined();
      context.setChildContext(childContext);
      expect(childContext.parentContext).toBe(context);
    });
  });

  describe('removeChildContext', () => {
    beforeEach(() => {
      resetContext();
      new Context({ suiteId: 'sanity' });
      new Context({});
    });

    test('sanity', () => {
      expect(singleton.useContext().suiteId).toBe('sanity');
      expect(singleton.useContext().parentContext.childContext).toBe(
        singleton.useContext()
      );
    });

    it('Should remove child context from context instance', () => {
      const parent = singleton.useContext().parentContext;
      parent.removeChildContext();
      expect(parent.childContext).toBeNull();
    });
  });

  describe('setParentContext', () => {
    let parent;
    beforeEach(() => {
      parent = new Context({});
      resetContext();
      new Context({});
    });

    it('Should set parent context to context instance', () => {
      expect(singleton.useContext().parentContext).toBeUndefined();
      singleton.useContext().setParentContext(parent);
      expect(singleton.useContext().parentContext).toBe(parent);
    });
  });

  describe('Nested context', () => {
    let context, parent1, parent2;

    beforeEach(() => {
      resetContext();
      parent1 = {
        prop1: 'parent_1_prop_1_value',
        prop2: 'parent_1_prop_2_value',
      };

      parent2 = {
        prop1: 'parent_2_prop_1_value',
        prop3: 'parent_2_prop_3_value',
      };
    });

    test('sanity', () => {
      expect(singleton.useContext()).toBeNull();
      context = new Context(parent1);
      expect(singleton.useContext()).toBe(context);
      expect(context).toMatchObject(parent1);
      expect(singleton.useContext().childContext).toBeUndefined();
      expect(singleton.useContext().parentContext).toBeUndefined();
    });

    it('Should match snapshot', () => {
      new Context({ suiteId: 10 });
      new Context(parent1);
      new Context(parent2);
      new Context(parent2);
      expect(singleton.useContext()).toMatchSnapshot();
    });

    it('Should add recent context as child of previous context', () => {
      new Context(parent1);
      new Context(parent2);
      expect(singleton.useContext()).toMatchObject(parent2);
      expect(singleton.useContext().parentContext).toMatchObject(parent1);
      expect(singleton.useContext().parentContext.childContext).toMatchObject(
        parent2
      );
      expect(singleton.useContext().parentContext.childContext).toBe(
        singleton.useContext()
      );
    });
  });
});
