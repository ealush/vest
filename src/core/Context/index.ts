import { singleton } from '../../lib';

/**
 * Creates a new context object, and assigns it as a static property on Vest's singleton.
 * @param {Object} parent   Parent context.
 */
function Context(parent) {
    singleton.use().ctx = this;
    Object.assign(this, parent);
}

/**
 * Sets a testObject reference on context.
 * @param {TestObject} A TestObject instance.
 */
Context.prototype.setCurrentTest = function(testObject) {
    this.currentTest = testObject;
};

/**
 * Removes current test from context.
 */
Context.prototype.removeCurrentTest = function() {
    delete this.currentTest;
};

/**
 * Clears stored instance from constructor function.
 */
Context.clear = function() {
    singleton.use().ctx = null;
};

export default Context;
