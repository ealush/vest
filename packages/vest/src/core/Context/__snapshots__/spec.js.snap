// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Nested context Should match snapshot 1`] = `
Context {
  "parentContext": Context {
    "childContext": [Circular],
    "parentContext": Context {
      "childContext": [Circular],
      "parentContext": Context {
        "childContext": [Circular],
        "suite_id": 10,
      },
      "prop1": "parent_1_prop_1_value",
      "prop2": "parent_1_prop_2_value",
    },
    "prop1": "parent_2_prop_1_value",
    "prop3": "parent_2_prop_3_value",
  },
  "prop1": "parent_2_prop_1_value",
  "prop3": "parent_2_prop_3_value",
}
`;
